src/groupByKey.ts

```diff
@@ -2,6 +2,43 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+interface Student {
+  name: string;
+  surname: string;
+  age: number;
+  married: boolean;
+  grades: number[];
+}
+
+interface Book {
+  author: string;
+  country: string;
+  imageLink: string;
+  language: string;
+  pages: number;
+  title: string;
+  year: number;
+}
+
+export function groupByKey(
+  items: [],
+  key: string,
+): GroupsMap<Book[] | Student[]> {
+  const values: string[] = [];
+
+  for (const item of items) {
+    const value = item[key];
+
+    if (!values.includes(value)) {
+      values.push(value);
+    }
```

Comment:
Storing unique values in an array and then filtering the original array for each unique value is inefficient, especially for large datasets. A more efficient approach would be to build the result object in a single iteration over the items without the need for the `values` array.